0006-operation-logging-and-monetary-resource-usage.md

```diff
@@ -0,0 +1,82 @@
+# Buildbarn Architecture Decision Record #6: Completed Action Logger and Monetary Resource Usage
+
+Authors: Travis Takai, Ed Schouten
+
+Date: 2021-03-23
+
+# Context
+
+Action results presented through bb-browser's frontend provide a great way to gain insight into the behavior of remote execution by looking at a single action at a time. The next logical step to this is analyzing build results in aggregate and in real-time. There are many reasons for wanting to analyze remote execution behavior such as identifying build duration trends, estimating the total cost of a build, identifying computationally expensive build targets, evaluating hardware utilization, or correlating completed build actions with data in the [Build Event Protocol](https://docs.bazel.build/versions/master/build-event-protocol.html). bb-browser displays all of the necessary info, but does not allow for programmatic analysis of action results for a number of reasons:
+
+* Most of the useful information is stored within the [ExecutedActionMetadata's auxiliary_metadata field](https://github.com/bazelbuild/remote-apis/blob/0943dc4e70e1414735a85a3167557392c177ff45/build/bazel/remote/execution/v2/remote_execution.proto#L945-L948) but is not exposed by the bazel client or displayed in the Build Event Protocol output.
+
+* There aren't any convenient ways for exploring build information via bb-browser as the instance name, action digest, and size of the action result are all needed before querying for an action result.
+
+* The goal of bb-storage is to offer short-term data storage, which does not allow for any form of querying for historical data.
+
+* bb-scheduler does not provide a persisted list of build actions and bb-storage has no way of providing access to a sequence of action results in an efficient way.
+
+An alternative approach that allows for a flexible configuration for clients is to allow for streaming action results along with their associated metadata to an external service. This allows for both long-term data persistence and real-time build action analysis. We should work towards creating this streaming service, which we can call Completed Action Logger.
+
+# Completed Action Logger Service
+
+The protocol will be used by bb_worker to stream details of completed actions, known as CompletedActions, to an external logging service:
+
+```protobuf
+service CompletedActionLogger {
+  // Send a CompletedAction to another service as soon as a build action has
+  // completed. Receiving a message from the return stream indicates that the
+  // client successfully received the CompletedAction.
+  rpc LogCompletedActions(stream CompletedAction)
+      returns (stream google.protobuf.Empty);
```

Comment:
Could you add a sentence here explaining what's the role of the return stream? For example:

> For every CompletedAction that the service processes successfully, an empty message is streamed back to the client.